helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -125,6 +211,79 @@ public void close() {
     _baseDataAccessor.close();
   }
 
+  @Override
+  public void handleDataChange(String dataPath, Object data) throws Exception {
+    Stat stat = new Stat();
+    ZNRecord readData =
+        _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+    LockInfo lockInfo = new LockInfo(readData);
+    // We are the current owner
+    if (lockInfo.getOwner().equals(_userId)) {
+      if (lockInfo.getRequestorId() == null || lockInfo.getPriority() > lockInfo
+          .getRequestorPriority()) {
+        LOG.info("We do not need to handle this data change");
+      } else {
+        _lockListener.onCleanupNotification();
+        // read the lock information again to avoid stale data
```

Comment:
Could you please elaborate on this part? I didn't get the necessity of reading again.
You can relies on the updater to check the ownership for you, right? If the owner has been changed in between, then the update fails and that is OK.

Reply:
The reason I was doing another read is that I wasn't using lock updater, but use a normal updater. I feel for these callback cases, we should not use lock updater, which is used during first tryLock. The reason is that we usually know what we are doing in the callback updater, and put all these logic inside lock updater will make that logic very complicated and hard to read. 
For example, in the lock updater, we always assume there is a normal lock request, which has owner ID, etc. However, the cleanup callback updater actually removes the owner ID. Do you think we should combine all scenarios in one updater?